clients/js/src/helpers/lifecycle.ts

```diff
@@ -28,23 +41,134 @@ export function canTransfer(
+  context: Pick<Context, 'eddsa' | 'rpc'>,
+  { authority, asset, collection, recipient }: ValidateTransferInput
+): Promise<null | string> {
+  const dAsset = deriveAssetPlugins(asset, collection);
+
+  // Permanent plugins have force approve powers
+  const permaTransferDelegate = checkPluginAuthorities({
+    authority,
+    pluginTypes: [PluginType.PermanentTransferDelegate],
+    asset: dAsset,
+    collection,
+  });
+  if (permaTransferDelegate.some((d) => d)) {
+    return null;
+  }
+
+  if (isFrozen(asset, collection)) {
+    return 'Unable to transfer: asset is frozen.';
+  }
+
+  if (dAsset.oracles?.length) {
+    const eligibleOracles = dAsset.oracles
+      .filter((o) =>
+        o.lifecycleChecks?.transfer?.includes(CheckResult.CAN_REJECT)
+      )
+      .filter((o) => {
+        // there's no PDA to derive, we can check the oracle account
+        if (!o.pda) {
+          return true;
+        }
+        // If there's a recipient in the inputs, we can try to check the oracle account
+        if (recipient) {
+          return true;
+        }
+
+        if (!getExtraAccountRequiredInputs(o.pda).includes('recipient')) {
+          return true;
+        }
+        // we skip the check if there's a recipient required but no recipient provided
+        // this is due how UIs generally show the availability of the transfer button before requiring the recipient address
+        return false;
+      });
+    if (eligibleOracles.length) {
+      const accountsWithOffset = eligibleOracles.map((o) => {
+        const account = findOracleAccount(context, o, {
+          asset: asset.publicKey,
+          collection: collection?.publicKey,
+          owner: asset.owner,
+          recipient,
+        });
+
+        return {
+          pubkey: account,
+          offset: o.resultsOffset,
+        };
+      });
+
+      const oracleValidations = (
+        await context.rpc.getAccounts(accountsWithOffset.map((a) => a.pubkey))
+      ).map((a, index) => {
+        if (a.exists) {
+          return deserializeOracleValidation(
+            a.data,
+            accountsWithOffset[index].offset
+          );
+        }
+        return null;
+      });
+
+      const oraclePass = oracleValidations.every(
+        (v) => v?.transfer === ExternalValidationResult.Pass
+      );
+      if (!oraclePass) {
+        return 'Unable to transfer: oracle validation failed.';
+      }
     }
-    const transferDelegates = checkPluginAuthorities({
-      authority,
-      pluginTypes: [PluginType.TransferDelegate],
-      asset: dAsset,
-      collection,
-    });
-    return transferDelegates.some((d) => d);
-  }
-  return false;
+  }
+
+  if (dAsset.owner === authority) {
+    return null;
+  }
+  const transferDelegates = checkPluginAuthorities({
+    authority,
+    pluginTypes: [PluginType.TransferDelegate],
+    asset: dAsset,
+    collection,
+  });
+  if (transferDelegates.some((d) => d)) {
+    return null;
+  }
```

Comment:
Can this method just call `canTransfer` to avoid duplication?

Same comment for `canBurn` / `validateBurn`, etc.

Reply:
not really because it returns a different error message based on why it failed

Reply:
Ok